frontend/src/components/Workspace/Visualize/VisualizeItemLayoutMenuIcon.tsx

```diff
@@ -10,13 +10,16 @@ import ListItemText from "@mui/material/ListItemText"
 import Menu from "@mui/material/Menu"
 import MenuItem from "@mui/material/MenuItem"
 
+import { cancelRoiApi } from "api/outputs/Outputs"
```

Comment:
@itutu-tienday I want to confirm this - Selecting a value in the Select ROI column and Select Items column will trigger onCancel(), correct? Or is onCancel() triggered only when selecting in the Select ROI column?

Reply:
@sanglevinh 
In all of the following cases, it is assumed that ROI Edit is Canceled.
1. select Items (-> Select Roi)
2. select Roi

The problem is that "the ROI edit mode has continued even though the user has moved from the current ROI edit screen".
If a move is made from the current ROI edit screen, the edit should basically be Canceled (discarded).

- Reference
  - https://docs.google.com/presentation/d/1Pz7N_bOCs2SuHq9RHHxZZHd79fnXhEHt/edit#slide=id.p2
  - If you are outside of this sequence, you are forced to Cancel (discard) any editing in progress.

*If I am not communicating it well, please let me know so I can add to it.

Reply:
@itutu-tienday I have edited that when the switch leaves the ROI (Visualization) edit screen, onCancel() will be executed, please check it for me.

Reply:
@sanglevinh 
Thank you.
Here is some feedback

- When calling the Cancel API, the following UI should also be returned to its pre-edit defaults.
![image](https://github.com/arayabrain/barebone-studio/assets/122338895/040ab366-bf5b-444f-8d9d-d2546f876667)

- "Select Roi" and "Add Roi" in combination several times caused the following FE error in some cases. Is this reproduced in your case?
```
ERROR
Cannot read properties of undefined (reading 'length')
TypeError: Cannot read properties of undefined (reading 'length')
    at addOrSelectRoi (http://localhost:3001/static/js/bundle.js:17130:37)
    at HTMLUnknownElement.callCallback (http://localhost:3001/static/js/bundle.js:400686:18)

```

Reply:
@itutu-tienday 
- I have fixed the issue of not returning to edit ROI when switching "Select ROI", 
- can you also help me check the ERROR you just said, I think it is related, if it is still there, can you describe it for me? elaborate on the difficult problem because I can't reproduce it

Reply:
@sanglevinh 
> I have fixed the issue of not returning to edit ROI when switching "Select ROI",

The above seems to be working. Thanks.

> can you also help me check the ERROR you just said,

The above does not seem to reproduce immediately, so I will let you know if there is a reproduction.

Sorry for the length of the thread, but I will provide additional feedback.

- Even when ROI Edit is not started, it seems that cancel API communication occurs when switching tabs, etc.
  - I believe that the above call to cancel API is unnecessary when ROI Edit is not started, but is it necessary?
  - *Please let us know if the state transition specification document needs to be more detailed.

- If "Select Roi" is not selected, "Edit ROI" menu should be hidden. 
  - Currently, "Edit ROI" is always displayed (operation stops in the middle).

Reply:
@itutu-tienday 
- It's true that calling the cancel api when switching tabs without editing the flag is not necessary, I have added a condition to do this.
- I also fixed the "Edit ROI" button to be displayed only when Select ROI is selected
Can you help me check to see if everything is okay?

Reply:
@sanglevinh 
Thank you.

I have checked the following operation patterns.
Overall, I think it is OK, but please check some NG items.

#### Edit ROI link display condition

  - [OK] When ROI is selected by "Select ROI"

#### Edit ROI Cancel (call in cancel API) timing

  - [NG] At the start of Edit ROI
  - [OK] During Edit ROI ...
    - [OK] When another item is selected in "Select Item"
    - [NG] When another ROI is selected in "Select Roi"
      - -> UI is not reset if data has not been updated yet.
    - [OK] When moving to another screen (another tab)
    - [OK] When reloading the browser
    - [OK] When exiting the browser

Reply:
@itutu-tienday 
I have fixed the problem
"[NG] When another ROI is selected in "Select Roi"
-> UI is not reset if data has not been updated.", pls check it again

but what about "[NG] When starting Edit ROI" you mean when you click Edit ROI it's called Cancel? If so, as far as I can see, cancel is not being called

Reply:
@sanglevinh 

> I have fixed the problem

confirmed fixed.

> but what about "[NG] When starting Edit ROI" you mean when you click Edit ROI it's called Cancel? If so, as far as I can see, cancel is not being called

Yes. Equivalent to "Initialize" in the following sequence diagram.
https://docs.google.com/presentation/d/1Pz7N_bOCs2SuHq9RHHxZZHd79fnXhEHt/edit#slide=id.p2
The purpose is to clear the cache data at the start of ROI Edit, because if the cancel API call fails, the cache data will remain.

Reply:
@itutu-tienday Sorry I don't have access, so I want to confirm that when pressing the "Edit ROI" button, the api cancel will always be called, right?

Reply:
@sanglevinh 
I see... I am attaching a sequence diagram.
![image](https://github.com/arayabrain/barebone-studio/assets/122338895/86a2f9db-f5b4-48e2-ad25-3e7fb26cdb55)

Note that I shared the specifications with the BE developer below, but I may not have shared the information properly.
https://github.com/arayabrain/barebone-studio/pull/107/files#r1400615777

Reply:
@itutu-tienday 
My understanding of this issue is as follows:
when in the Visualize screen
 (1) -> select Item
 (2) -> select ROI
then onCancel() will be called right?
If my understanding is correct but you don't see onCancel() being called then it's because I'm checking statusRoi. If statusRoi is there, I will call onCancel().

I want to hear your opinion.
Sorry if I misunderstood you

Reply:
@sanglevinh 
Yes, it seems that the flow you described would be fine. (It meets the requirements of the sequence diagram)

Therefore, I would like to conclude the issue of the call pattern of the cancel API.

As a supplementary note, I have confirmed that the following operation pattern brings up the cached data that was previously edited.
1. force close of the browser (cancel API is not called)
2. Start the browser again and display the same ROI on the Visualize screen → ROI Edit mode remains.
If the above does not cause any particular problem, I consider it OK as it is now.

Reply:
@itutu-tienday 
about this problem, I changed the job to only call Api cancel when
 - press Cancel,
 - Press the button
Close (X)
 - switch tab (e.g. Visualize -> Record)

Reply:
@sanglevinh 
The above conditions look good. Thanks.